File: Ejemplos/43-DockerAndTestContainers/src/test/java/dev/joseluisgs/services/PersonasServiceImplTest.java
package dev.joseluisgs.services;

import dev.joseluisgs.errors.PersonaError;
import dev.joseluisgs.errors.PersonaNotFound;
import dev.joseluisgs.model.Persona;
import dev.joseluisgs.repository.PersonasRepository;
import io.vavr.control.Either;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class PersonasServiceImplTest {

    @Mock
    private PersonasRepository personasRepository;

    @InjectMocks
    private PersonasServiceImpl personasService;


    @Test
    void getAllSuccess() {
        List<Persona> personas = List.of(Persona.builder().id(1).nombre("John Doe").build());
        when(personasRepository.getAll()).thenReturn(personas);

        Either<PersonaError, List<Persona>> result = personasService.getAll();

        assertAll(() -> {
            assertTrue(result.isRight());
            assertEquals(personas, result.get());

        });

        verify(personasRepository, times(1)).getAll();
    }

    @Test
    void getByIdSuccess() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.getById(1)).thenReturn(Optional.of(persona));

        Either<PersonaError, Persona> result = personasService.getById(1);

        assertAll(() -> {
            assertTrue(result.isRight());
            assertEquals(persona, result.get());
        });
        verify(personasRepository, times(1)).getById(1);
    }

    @Test
    void getByIdNotFound() {
        when(personasRepository.getById(1)).thenReturn(Optional.empty());

        Either<PersonaError, Persona> result = personasService.getById(1);

        assertAll(() -> {
            assertTrue(result.isLeft());
            assertInstanceOf(PersonaNotFound.class, result.getLeft());
            assertEquals("Persona no encontrada con id: 1", result.getLeft().getMessage());
        });

        verify(personasRepository, times(1)).getById(1);
    }

    @Test
    void createSuccess() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.create(persona)).thenReturn(persona);

        Either<PersonaError, Persona> result = personasService.create(persona);

        assertAll(() -> {
            assertTrue(result.isRight());
            assertEquals(persona, result.get());
        });

        verify(personasRepository, times(1)).create(persona);
    }

    @Test
    void createError() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.create(persona)).thenThrow(new RuntimeException());

        Either<PersonaError, Persona> result = personasService.create(persona);

        assertTrue(result.isLeft());
        assertInstanceOf(PersonaError.class, result.getLeft());
        verify(personasRepository, times(1)).create(persona);
    }

    @Test
    void updateSuccess() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.getById(1)).thenReturn(Optional.of(persona));
        when(personasRepository.update(1, persona)).thenReturn(persona);

        Either<PersonaError, Persona> result = personasService.update(1, persona);

        assertAll(() -> {
            assertTrue(result.isRight());
            assertEquals(persona, result.get());
        });

        assertTrue(result.isRight());
        assertEquals(persona, result.get());


        verify(personasRepository, times(1)).getById(1);
        verify(personasRepository, times(1)).update(1, persona);
    }

    @Test
    void updateNotFound() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.getById(1)).thenReturn(Optional.empty());

        Either<PersonaError, Persona> result = personasService.update(1, persona);

        assertAll(() -> {
            assertTrue(result.isLeft());
            assertInstanceOf(PersonaNotFound.class, result.getLeft());
            assertEquals("Persona no encontrada con id: 1", result.getLeft().getMessage());
        });

        verify(personasRepository, times(1)).getById(1);
        verify(personasRepository, never()).update(1, persona);
    }

    @Test
    void deleteSuccess() {
        Persona persona = Persona.builder().id(1).nombre("John Doe").build();
        when(personasRepository.getById(1)).thenReturn(Optional.of(persona));

        Either<PersonaError, Persona> result = personasService.delete(1);

        assertAll(() -> {
            assertTrue(result.isRight());
            assertEquals(persona, result.get());
        });

        verify(personasRepository, times(1)).getById(1);
        verify(personasRepository, times(1)).delete(1);
    }

    @Test
    void deleteNotFound() {
        when(personasRepository.getById(1)).thenReturn(Optional.empty());

        Either<PersonaError, Persona> result = personasService.delete(1);

        assertAll(() -> {
            assertTrue(result.isLeft());
            assertInstanceOf(PersonaNotFound.class, result.getLeft());
            assertEquals("Persona no encontrada con id: 1", result.getLeft().getMessage());
        });
        
        verify(personasRepository, times(1)).getById(1);
        verify(personasRepository, never()).delete(1);
    }
}

File: Ejemplos/33-EjemplosTestRepositorioServicio/src/test/kotlin/personas/services/PersonasServiceImplTest.kt
package personas.services

import dev.joseluisgs.personas.exceptions.PersonaException
import dev.joseluisgs.personas.model.Persona
import dev.joseluisgs.personas.repository.PersonasRepository
import dev.joseluisgs.personas.services.PersonasServiceImpl
import dev.joseluisgs.personas.services.cache.CachePersonas
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@ExtendWith(MockKExtension::class)
class PersonasServiceImplTest {

    @MockK
    private lateinit var cache: CachePersonas

    @MockK
    private lateinit var repository: PersonasRepository

    @InjectMockKs
    private lateinit var personasService: PersonasServiceImpl

    private val persona1 = Persona(UUID.fromString("63591beb-d620-4dcf-a719-3a2c4ed88f02"), "Test1", 18)
    private val persona2 = Persona(UUID.fromString("63591beb-d620-4dcf-a719-3a2c4ed88f03"), "Test2", 19)
    private val personas = listOf(persona1, persona2)

    @BeforeEach
    fun setUp() {
    }

    @Test
    fun getAll() {
        // Arrange
        every { repository.getAll() } returns personas

        // Act
        val result = personasService.getAll()

        // Assert
        assertAll(
            { assert(result.isNotEmpty()) },
            { assert(result.size == 2) },
            { assert(result.contains(persona1)) },
            { assert(result.contains(persona2)) }
        )

        // Verify
        verify(atLeast = 1) { repository.getAll() }
    }

    @Test
    fun findByName() {
        // Arrange
        every { repository.findByName("Test1") } returns listOf(persona1)

        // Act
        val result = personasService.findByName("Test1")

        // Assert
        assertAll(
            { assert(result.isNotEmpty()) },
            { assert(result.size == 1) },
            { assert(result.contains(persona1)) }
        )

        // Verify
        verify(atLeast = 1) { repository.findByName("Test1") }
    }

    @Test
    fun getByIdInCache() {
        // Arrange
        every { cache.get(persona1.ID) } returns persona1

        // Act
        val result = personasService.getById(persona1.ID)

        // Assert
        assertAll(
            { assert(result == persona1) },
            { assert(result.ID == persona1.ID) },
            { assert(result.nombre == persona1.nombre) },
            { assert(result.edad == persona1.edad) }
        )

        // Verify
        verify(atLeast = 0) { repository.getById(persona1.ID) }
        verify(atLeast = 1) { cache.get(persona1.ID) }
    }

    @Test
    fun getByIdNotFoundInCacheButInRepository() {
        // Arrange
        every { cache.get(persona1.ID) } returns null
        every { repository.getById(persona1.ID) } returns persona1
        every { cache.put(persona1.ID, persona1) } returns Unit

        // Act
        val result = personasService.getById(persona1.ID)

        // Assert
        assertAll(
            { assert(result == persona1) },
            { assert(result.ID == persona1.ID) },
            { assert(result.nombre == persona1.nombre) },
            { assert(result.edad == persona1.edad) }
        )

        // Verify
        verify(atLeast = 1) { repository.getById(persona1.ID) }
        verify(atLeast = 1) { cache.get(persona1.ID) }
    }

    @Test
    fun getByIdNotFoundInCacheAndRepository() {
        // Arrange
        every { cache.get(persona1.ID) } returns null
        every { repository.getById(persona1.ID) } returns null

        // Act
        val message = assertThrows<PersonaException.PersonaNotFoundException> {
            personasService.getById(persona1.ID)
        }

        // Assert
        assertAll(
            { assert(message.message == "Persona no encontrada con ID: ${persona1.ID}") }
        )

        // Verify
        verify(atLeast = 1) { repository.getById(persona1.ID) }
        verify(atLeast = 1) { cache.get(persona1.ID) }
    }

    @Test
    fun save() {
        // Arrange
        every { repository.save(persona1) } returns persona1

        // Act
        val result = personasService.save(persona1)

        // Assert
        assertAll(
            { assert(result.ID == persona1.ID) },
            { assert(result.nombre == persona1.nombre) },
            { assert(result.edad == persona1.edad) }
        )

        // Verify
        verify(atLeast = 1) { repository.save(persona1) }
        verify(atLeast = 0) { cache.put(persona1.ID, persona1) }
    }

    @Test
    fun update() {
        // Arrange
        every { repository.update(persona1.ID, persona1) } returns persona1
        every { cache.remove(persona1.ID) } returns Unit

        // Act
        val result = personasService.update(persona1.ID, persona1)

        // Assert
        assertAll(
            { assert(result.ID == persona1.ID) },
            { assert(result.nombre == persona1.nombre) },
            { assert(result.edad == persona1.edad) }
        )

        // Verify
        verify(atLeast = 1) { repository.update(persona1.ID, persona1) }
        verify(atLeast = 1) { cache.remove(persona1.ID) }
    }

    @Test
    fun updateNotFound() {
        // Arrange
        every { repository.update(persona1.ID, persona1) } returns null
        every { cache.remove(persona1.ID) } returns Unit

        // Act
        val message = assertThrows<PersonaException.PersonaNotUpdatedException> {
            personasService.update(persona1.ID, persona1)
        }

        // Assert
        assertAll(
            { assert(message.message == "Persona no se ha podido actualizar con ID: ${persona1.ID}") }
        )

        // Verify
        verify(atLeast = 1) { repository.update(persona1.ID, persona1) }
        verify(atLeast = 1) { cache.remove(persona1.ID) }
    }

    @Test
    fun delete() {
        // Arrange
        every { repository.delete(persona1.ID) } returns persona1
        every { cache.remove(persona1.ID) } returns Unit

        // Act
        val result = personasService.delete(persona1.ID)

        // Assert
        assertAll(
            { assert(result.ID == persona1.ID) },
            { assert(result.nombre == persona1.nombre) },
            { assert(result.edad == persona1.edad) }
        )

        // Verify
        verify(atLeast = 1) { repository.delete(persona1.ID) }
        verify(atLeast = 1) { cache.remove(persona1.ID) }
    }

    @Test
    fun deleteNotFound() {
        // Arrange
        every { repository.delete(persona1.ID) } returns null
        every { cache.remove(persona1.ID) } returns Unit

        // Act
        val message = assertThrows<PersonaException.PersonaNotDeletedException> {
            personasService.delete(persona1.ID)
        }

        // Assert
        assertAll(
            { assert(message.message == "Persona no se ha podido eliminar con ID: ${persona1.ID}") }
        )

        // Verify
        verify(atLeast = 1) { repository.delete(persona1.ID) }
        verify(atLeast = 1) { cache.remove(persona1.ID) }
    }
}